### PR TITLE
Do not create backend_api automatically when create a Service

### PR DESCRIPTION
**Pull request by duduribeiro**
Closes [THREESCALE-3368](https://issues.jboss.org/browse/THREESCALE-3368
)

This PR removes the auto creation of a backend api when a
product is created. We don't need to have a backend api every time
that a product is created.

The idea is that product and backend are two different things that
can be independently created and linked together at will.
All services are/will be products and there will then be no need
for a checkbox.

When using RU APIAP, the private endpoint is not created.
It creates a Service with default api_backend when RU APIAP is disabled

### CODE REVIEW

File: app/models/create_service_with_backend_api.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+#   CreateServiceWithBackendApi.new(service: service, account: account, backend_api: 1)
+#   # => true # This will only attach the service with id 1 to the service.
+class CreateServiceWithBackendApi
+  def initialize(service:, account:, backend_api:)
+    @account     = account
+    @backend_api = backend_api
+    @service     = service
+  end
+
+  def save
```
Comment (hallelujah) [inline app/models/create_service_with_backend_api.rb]: `def call` though `#save` works too
It is just convention

File: app/controllers/api/services_controller.rb [hunk 2]
```diff
@@ -37,9 +37,9 @@ def settings
   def create
     @service = collection.new # this is done in 2 steps so that the account_id is in place as preffix_key relies on it
     @service.attributes = service_params
```
Comment (hallelujah) [inline app/controllers/api/services_controller.rb]: Those should be set in the CreateService object

File: app/controllers/api/services_controller.rb [hunk 3]
```diff
@@ -37,9 +37,9 @@ def settings
   def create
     @service = collection.new # this is done in 2 steps so that the account_id is in place as preffix_key relies on it
```
Comment (hallelujah) [inline app/controllers/api/services_controller.rb]: Does `current_user.accessible_services.new` add anything to `@account.services.new`?
If not then it can be done in the service object

File: app/controllers/api/services_controller.rb (hunk 2 above)
Comment (Martouta) [reply app/controllers/api/services_controller.rb]: @hallelujah @duduribeiro Ignore this for now plz because it is fixed in my PR and u would make me have more conflicts unnecessarily if u change it also here 😝

File: app/controllers/api/services_controller.rb (hunk 3 above)
Comment (Martouta) [reply app/controllers/api/services_controller.rb]: Same as https://github.com/3scale/porta/pull/1168#discussion_r322262326
This is not a simple topic so leave it for my PR plz

Comment (hallelujah) [reply app/controllers/api/services_controller.rb]: okidoki :)

File: app/controllers/admin/api/services_controller.rb [hunk 4]
```diff
@@ -42,11 +42,14 @@ def index
   ##~ op.parameters.add :name => "description", :description => "Description of the service to be created.", :dataType => "string", :required => false, :paramType => "query"
   ##~ op.parameters.add :name => "deployment_option", :description => "Deployment option for the gateway: 'hosted' for APIcast hosted, 'self_managed' for APIcast Self-managed option", :dataType => "string", :allowMultiple => false, :required => false, :paramType => "query"
   ##~ op.parameters.add :name => "backend_version", :description => "Authentication mode: '1' for API key, '2' for App Id / App Key, 'oidc' for OpenID Connect", :dataType => "string", :allowMultiple => false, :required => false, :paramType => "query"
+  ##~ op.parameters.add :name => "backend_api_id", :description => "ID of the backend api that will be associated with the service", :dataType => "string", :allowMultiple => false, :required => false, :paramType => "query"
```
Comment (Martouta) [inline app/controllers/admin/api/services_controller.rb]: I think that we shouldn't document it until it is available for everyone 🤔 WDYT @hallelujah ?

File: test/integration/api/services_controller_test.rb [hunk 5]
```diff
@@ -82,43 +82,57 @@ def update_params
     end
   end
 
-  class ActAsProduct < Api::ServicesControllerTest
+  class BackendApiSelection < Api::ServicesControllerTest
```
Comment (Martouta) [inline test/integration/api/services_controller_test.rb]: We should add these tests for the API side as well 😄

File: app/controllers/admin/api/services_controller.rb (hunk 4 above)
Comment (hallelujah) [reply app/controllers/admin/api/services_controller.rb]: We should not you are right
All using API endpoints. There is a workflow that is broken, as it is with this removal:

1. Create a Service without `api_backend` parameter
2. `Proxy Config Promote
/admin/api/services/{service_id}/proxy/configs/{environment}/{version}/promote.json`
3. We should not be able to deploy as `api_backend` (private_endpoint) is blank

The customer should update either Service with api_backend or Proxy with private_endpoint or api_backend

This should be transitional state.

So this is broken for quick service creation, as it needs another UPDATE to deploy the proxy

File: gemfiles/prod/Gemfile.lock [hunk 6]
```diff
@@ -513,7 +513,6 @@ GEM
     pry (0.12.2)
       coderay (~> 1.1.0)
       method_source (~> 0.9.0)
-      slop (~> 3.4)
```
Comment (Martouta) [inline gemfiles/prod/Gemfile.lock]: 👀

File: app/lib/backend_api_logic/service_extension.rb [hunk 7]
```diff
@@ -18,27 +18,24 @@ module ServiceExtension
...
     def first_backend_api
-      @backend_api ||= find_or_create_first_backend_api!
+      @backend_api ||= find_first_backend_api!
     end
 
-    def find_or_create_first_backend_api!
-      return unless account
-      config = backend_api_configs.first || create_first_backend_api_config!
-      config.backend_api
+    def find_first_backend_api!
```
Comment (Martouta) [inline app/lib/backend_api_logic/service_extension.rb]: This shouldn't be a bang method anymore considering its content is `backend_api_configs.first&.backend_api` 😄 
```suggestion
    def find_first_backend_api
```

Comment (Martouta) [reply app/lib/backend_api_logic/service_extension.rb]: In fact we don't even need this method 🤔 we could do in the method above:
```
def first_backend_api
  @backend_api ||= backend_api_configs.first&.backend_api
```

File: app/lib/backend_api_logic/service_extension.rb [hunk 8]
```diff
@@ -18,27 +18,24 @@ module ServiceExtension
     end
 
     def first_backend_api
-      @backend_api ||= find_or_create_first_backend_api!
+      @backend_api ||= find_first_backend_api!
```
Comment (Martouta) [inline app/lib/backend_api_logic/service_extension.rb]: https://github.com/3scale/porta/pull/1168/files#r324695146

File: config/webpacker.yml [hunk 9]
```diff
@@ -5,6 +5,7 @@ default: &default
   source_entry_path: packs
   public_output_path: packs
   cache_path: tmp/cache/webpacker
+  webpack_compile_output: true
```
Comment (Martouta) [inline config/webpacker.yml]: Is this related to this PR? 😱 What for? 🤔

Comment (duduribeiro) [reply config/webpacker.yml]: I added it because it is easier to debug when something fails in the JS.. but It shouldn't be here in the PR 😄 

will remove it

**Comment by duduribeiro** [comment]
@hallelujah @guicassolato @Martouta 

I will follow to not promotion of api when it has 0 backends, like suggested here: https://3scale.slack.com/archives/CLMKJCNKF/p1568405174223600

Should we follow this in this PR or in a new one? 🤔

**Comment by hallelujah** [comment]
@duduribeiro in another PR.
Small changes are fine and this one does not really break anything right?
The proxy config will just have `api_backend` set to `null` in the JSON

File: app/controllers/admin/api/services_controller.rb [hunk 10]
```diff
@@ -46,7 +46,9 @@ def index
   ##~ op.parameters.add :name => " ", :dataType => "custom", :paramType => "query", :allowMultiple => true, :description => "Extra parameters"
   #
   def create
-    service = current_account.create_service(service_params)
+    service = current_account.accessible_services.new(service_params)
+    create_service = CreateServiceWithBackendApi.new(service: service, backend_api_id: params[:backend_api_id].presence)
```
Comment (hallelujah) [inline app/controllers/admin/api/services_controller.rb]: If `backend_api_id` is an optional parameter I will call it `CreateServiceProcedure` (because CreateServiceService is ugly)

File: app/lib/backend_api_logic/service_extension.rb [hunk 11]
```diff
@@ -18,27 +18,20 @@ module ServiceExtension
...
-    def find_or_create_first_backend_api!
-      return unless account
-      config = backend_api_configs.first || create_first_backend_api_config!
-      config.backend_api
-    end
-
-    private
-
-    def create_first_backend_api_config!
+    def build_default_backend_api_config(endpoint = nil)
```
Comment (guicassolato) [inline app/lib/backend_api_logic/service_extension.rb]: This still looks like `build_or_create_default_backend_api`.

File: app/models/proxy.rb [hunk 12]
```diff
@@ -362,7 +362,9 @@ def authorization_credentials
   end
 
   def skip_test_request?
-    service.oauth? || saas_configuration_driven_apicast_self_managed?
+    # !backend_api || service.oauth? || saas_configuration_driven_apicast_self_managed? This should be improved like
+    # pointed here: https://github.com/3scale/porta/pull/1168#discussion_r323736488
+    return true
```
Comment (guicassolato) [inline app/models/proxy.rb]: I think removing this test does not belong to this PR.

File: app/services/create_service_with_backend_api.rb [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+  end
+
+  def call
+    build_backend_api if @account.provider_can_use?(:api_as_product)
+    @service.save
+  end
+
+  private
+
+  def build_backend_api
```
Comment (guicassolato) [inline app/services/create_service_with_backend_api.rb]:
```suggestion
  def build_backend_api_config
```

File: app/services/create_service_with_backend_api.rb [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+    return unless @backend_api_id
+
+    backend_api = @account.backend_apis.find(@backend_api_id)
+    @service.backend_api_configs = [BackendApiConfig.new(backend_api: backend_api)]
```
Comment (guicassolato) [inline app/services/create_service_with_backend_api.rb]:
```suggestion
    @service.backend_api_configs.build(backend_api: backend_api)
```

Considering the name of the method, I'd expect the new backend api config to be returned and not the collection anyway.

File: features/old/menu/dashboard.feature [hunk 15]
```diff
@@ -1,10 +1,10 @@
 Feature: Dashboard
   In order to navigate easily
-  As a provider
+  As a provider with backend api
   I want to have important links on the dashboard
 
   Background:
-    Given a provider "foo.example.com"
+    Given a provider "foo.example.com" with backend api
```
Comment (guicassolato) [inline features/old/menu/dashboard.feature]: "Given a provider" implies "with a service". Why wouldn't it imply "with backend api"? I think we should define tests in the other way around: "Given a provider **without** backend api" and avoid changing a bunch of cukes.

File: spec/acceptance/api/proxy_spec.rb [hunk 16]
```diff
@@ -6,6 +6,10 @@
 
   let(:service_id) { service.id }
 
+  before do
+    service.build_default_backend_api_config.save!
```
Comment (guicassolato) [inline spec/acceptance/api/proxy_spec.rb]: I don't get it. Why `save!` on the backend API if the the method already follows the state of the service either calling `.new` or `.create!`?

File: app/lib/backend_api_logic/service_extension.rb (hunk 11 above)
Comment (guicassolato) [reply app/lib/backend_api_logic/service_extension.rb]: Renaming this method will also break https://github.com/3scale/porta/blob/8cecd33c61eed6e724ffcc2108effb89fd913ac0/db/migrate/20190716110520_create_the_backend_apis_of_services.rb#L14

File: app/lib/backend_api_logic/service_extension.rb [hunk 17]
```diff
@@ -18,27 +18,20 @@ module ServiceExtension
...
-        private_endpoint: proxy&.[]('api_backend')
+        private_endpoint: endpoint || proxy&.[]('api_backend')
       )
       constructor = new_record? ? :build : :create!
       attrs = {backend_api: backend_api, path: ''}
       backend_api_configs.public_send(constructor, attrs)
     end
   end
-end
+end
```
Comment (Martouta) [inline app/lib/backend_api_logic/service_extension.rb]: 👀

File: app/services/create_service_with_backend_api.rb [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+  end
+end
```
Comment (Martouta) [inline app/services/create_service_with_backend_api.rb]: 👀

File: spec/acceptance/api/proxy_spec.rb (hunk 16 above)
Comment (Martouta) [reply spec/acceptance/api/proxy_spec.rb]: For this: https://github.com/3scale/porta/blob/679f351d2382e677f557c602ad73b7122c905efd/spec/acceptance/api/proxy_spec.rb#L31

This does not follow any api call.

File: test/functional/api/integrations_controller_test.rb [hunk 19]
```diff
@@ -5,6 +5,7 @@ class Api::IntegrationsControllerTest < ActionController::TestCase
   def setup
     @provider = FactoryBot.create(:provider_account)
     @provider.default_service.service_tokens.create!(value: 'token')
+    @provider.default_service.build_default_backend_api_config.save!
```
Comment (Martouta) [inline test/functional/api/integrations_controller_test.rb]: What for? funcional tests are deprecated and we mean to remove them bit by bit

File: test/functional/provider/admin/onboarding/wizard/request_controller_test.rb [hunk 20]
```diff
@@ -4,6 +4,7 @@ class Provider::Admin::Onboarding::Wizard::RequestControllerTest < ActionControl
 
   def setup
     stub_request(:get, %r{http://test\.proxy/deploy}).to_return(status: 200)
+    master_account.first_service.build_default_backend_api_config.save!
```
Comment (Martouta) [inline test/functional/provider/admin/onboarding/wizard/request_controller_test.rb]: Same as https://github.com/3scale/porta/pull/1168/files#r325407393

File: app/controllers/admin/api/services_controller.rb [hunk 21]
```diff
@@ -46,7 +46,9 @@ def index
   ##~ op.parameters.add :name => " ", :dataType => "custom", :paramType => "query", :allowMultiple => true, :description => "Extra parameters"
   #
   def create
-    service = current_account.create_service(service_params)
+    service = current_account.accessible_services.new(service_params)
```
Comment (Martouta) [inline app/controllers/admin/api/services_controller.rb]:
```suggestion
    service = current_account.services.new(service_params)
```

It is going to be accessible anyway so let's don't complicate our lifes 😄

File: app/controllers/api/integrations_controller.rb [hunk 22]
```diff
@@ -30,13 +30,14 @@ def update
       proxy_pro_update
     elsif @proxy.save_and_deploy(proxy_params)
       environment = @proxy.service_mesh_integration? ? 'Production' : 'Staging'
+      api_backend = @proxy.api_backend
```
Comment (Martouta) [inline app/controllers/api/integrations_controller.rb]: Minor comment but imo better if this goes in the line after the `if @proxy.send_api_test_request!` because if it does not reach there, this is not needed so better not to execute stuff that is not needed

File: app/controllers/api/services_controller.rb [hunk 23]
```diff
@@ -69,15 +69,14 @@ def destroy
...
-    return [] if !provider_can_use?(:api_as_product) || params[:service][:backend_api].blank?
-    backend_api = current_account.backend_apis.find(params[:service][:backend_api])
-    [BackendApiConfig.new(backend_api: backend_api)]
-  end
-
   def service_params
-    allowed_params = %i[system_name name description]
-    params.require(:service).permit(allowed_params)
+    permitted_params = [:name, :system_name, :description, :support_email, :deployment_option, :backend_version,
+      :intentions_required, :buyers_manage_apps, :referrer_filters_required,
```
Comment (Martouta) [inline app/controllers/api/services_controller.rb]: 😝

image

File: Gemfile.base [hunk 24]
```diff
@@ -88,6 +88,7 @@ gem '3scale_client', '~> 2.6.1', require: false
 gem 'analytics-ruby', require: false
 
 group :development, :test do
+  gem 'active_record_query_trace'
```
Comment (duduribeiro) [inline Gemfile.base]: all this gems are because it is wip.. will remove it when I finish

File: app/controllers/admin/api/services_controller.rb [hunk 25]
```diff
@@ -46,7 +46,9 @@ def index
   ##~ op.parameters.add :name => " ", :dataType => "custom", :paramType => "query", :allowMultiple => true, :description => "Extra parameters"
   #
   def create
-    service = current_account.create_service(service_params)
+    service = current_account.accessible_services.new
```
Comment (hallelujah) [inline app/controllers/admin/api/services_controller.rb]: well using the scope does not change much so I'd say
```suggestion
    service = current_account.build_service
```

File: app/controllers/api/integrations_controller.rb [hunk 26]
```diff
@@ -36,7 +37,7 @@ def update
 
       if @proxy.send_api_test_request!
         onboarding.bubble_update('api')
-        done_step(:api_sandbox_traffic) if ApiClassificationService.test(@proxy.api_backend).real_api?
+        done_step(:api_sandbox_traffic) if @proxy.api_backend && ApiClassificationService.test(@proxy.api_backend).real_api?
```
Comment (hallelujah) [inline app/controllers/api/integrations_controller.rb]: `@proxy.api_backend.present?`

File: app/controllers/api/services_controller.rb [hunk 27]
```diff
@@ -36,10 +36,9 @@ def settings
 
   def create
     @service = collection.new # this is done in 2 steps so that the account_id is in place as preffix_key relies on it
-    @service.attributes = service_params
-    @service.backend_api_configs = build_backend_api_configs
+    creator = ServiceCreator.new(service: @service, backend_api: find_backend_api)
```
Comment (hallelujah) [inline app/controllers/api/services_controller.rb]: We decided that we should remove the Backend API selector when creating a Service in the UI

File: app/controllers/api/services_controller.rb [hunk 28]
```diff
@@ -69,17 +68,28 @@ def destroy
...
-  end
-
   def service_params
     allowed_params = %i[system_name name description]
     params.require(:service).permit(allowed_params)
   end
 
+  private
+
+  def find_backend_api
```
Comment (hallelujah) [inline app/controllers/api/services_controller.rb]: No need of it, the service will always be created without a BackendApi

File: app/controllers/api/integrations_controller.rb (hunk 26 above)
Comment (Martouta) [reply app/controllers/api/integrations_controller.rb]: we shouldn't be calling `@proxy.api_backend` twice anyway 😄 better to assign it to a variable

File: app/lib/backend_api_logic/service_extension.rb [hunk 29]
```diff
@@ -17,28 +19,38 @@ module ServiceExtension
...
+      end
+
+      def backend_api_config
+        @backend_api_config ||= backend_api.backend_api_configs.first ||
+                                backend_api_configs.build(path: '', backend_api: backend_api)
+      end
+
+      def backend_api
+        return unless account
+        @backend_api ||= backend_api_configs.first&.backend_api || account.backend_apis.build(name: service.system_name, name: "#{service.name} Backend API", description: "Backend API of #{service.name}")
```
Comment (hallelujah) [inline app/lib/backend_api_logic/service_extension.rb]: So under RU, the workflow is for API endpoint and UI

1. Create a Backend API
2. Create a Service
3 Link the backend API to this service

When the RU is off

1. Create the Service, the backend api is there by default

So we need to change that there too

File: app/forms/onboarding/request_form.rb [hunk 30]
```diff
@@ -33,6 +33,7 @@ def proxy_base_url
   end
 
   def uri
+    return unless model.api_backend
```
[the remark this answers is not in the document]
Comment (duduribeiro) [reply app/forms/onboarding/request_form.rb]: I added this because it is possible to not have a private_endpoint. If it does not have, it fails trying to create an URI

[the remark this answers is not in the document]
Comment (hallelujah) [reply app/forms/onboarding/request_form.rb]: OK

File: app/forms/onboarding/api_form.rb [hunk 31]
```diff
@@ -29,6 +29,14 @@ def example_backend
     ECHO_API_BACKEND
   end
 
+  def save
```
Comment (hallelujah) [inline app/forms/onboarding/api_form.rb]: I do not think we need this

The onboarding does not create a service, it just updates the ones created in ProviderSignup#signup_provider

File: test/factories/account.rb [hunk 32]
```diff
@@ -188,6 +190,13 @@
     end
   end
 
+  factory(:provider_account_with_default_backend_api, parent: :provider_account) do
```
Comment (hallelujah) [inline test/factories/account.rb]: This could be a trait

File: app/forms/onboarding/api_form.rb (hunk 31 above)
Comment (duduribeiro) [reply app/forms/onboarding/api_form.rb]: Makes total sense

**Comment by duduribeiro** [comment]
Closed because it is addressed in https://github.com/3scale/porta/pull/1211
